php/libraries/CandidateData.class.inc

```diff
@@ -0,0 +1,88 @@
+<?php declare(strict_types=1);
+/**
+ * A CandidateData object is a DTO (data transfer object) to
+ * store a representation of data associated with a Candidate
+ * object in LORIS.
+ *
+ * The CandidateData may be partially loaded, in which case any
+ * getter trying to access an unknown will throw an exception.
+ * This allows modules to only load relevant data in a query,
+ * and avoid the overhead of the Candidate::singleton() constructor.
+ *
+ * For instance, a query that selected CenterID and ProjectID and
+ * timepoint related information:
+ * the database could instantiate a Candidate as:
+ *    $visit = new Candidate(new CandidateData($projectid, $centerid, $timepoints));
+ *
+ * with data loaded from the query, and then call
+ * `$candidate->isAccessibleBy($user)` without having to load all the
+ * other data that Candidate::singleton() does. If an attempt is
+ * made to access a property not in the DTO, an exception is thrown,
+ * rather than silently returning the incorrect value.
+ *
+ * @license http://www.gnu.org/licenses/gpl-3.0.txt GPLv3
+ */
+class CandidateData
+{
+    protected $registrationProjectID;
```

Comment:
I would type these variables, and make them public. You can then get rid of all you getters I believe because unless they are initialized, they will throw an error if accessed

Reply:
No, that doesn't work.. then you need to pass all of them and there's no way to partially construct the object

Reply:
you can set them one by one since they are public. if you want the constructor, you can set if statements that dont set the variable if its passed a null value and then set it manually ?

Reply:
done